### PR TITLE
feat(logging): OpenTelemetry trace/span ID integration for Java logging library

### LINKED ISSUE #1595: Java-Logging: OpenTelemetry spans are ignored as parents for LogRecords
Tracking the same issue as https://github.com/googleapis/google-cloud-go/issues/9302 to propagate correct TraceId and SpanId for logging and tracing correlation.

**Client**
[java-logging](https://github.com/googleapis/java-logging)

**Context**
In java-logging libraries, the current implementation will always use a remote trace ID and span ID parsed from HTTP headers that originate from outside the application.

**Expected behavior**
If the user is using opentelemetry instrumentation, we want to use the traceID and spanID of the span that OpenTelemetry created. This way, when a user looks for logs associated with that span, the logs produced by this logging library will appear.

**Design Docs:**
[OpenTelemetry Trace/Span ID Support in Cloud Logging Library](http://goto.google.com/logging-otel-span-support)
[OpenTelemetry Trace/Span ID Support in Java Cloud Logging Library](http://go/logging-otel-span-support-java)

### PR DESCRIPTION
**Pull request by cindy-peng**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-logging/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #1595  ☕️

This change is to add OpenTelemetry support for java-logging library and fix trace-span correlation issue.
More context about this change: https://b.corp.google.com/issues/270985947

Here are the overall and java-logging design doc for this change:
[go/logging-otel-span-support](http://goto.google.com/logging-otel-span-support)
[go/logging-otel-span-support-java](http://go/logging-otel-span-support-java)

### CODE REVIEW

**Review by bhshkh — APPROVED** [review]

File: google-cloud-logging/src/test/java/com/google/cloud/logging/ContextTest.java [hunk 1]
```diff
@@ -114,40 +127,78 @@ public void testParsingCloudTraceContext() {
...
-    assertTraceAndSpan(builder.build(), null, null);
+    assertTraceSpanAndSampled(builder.build(), null, null, false);
     builder.loadW3CTraceParentContext(W3C_TRACE_CONTEXT);
-    assertTraceAndSpan(builder.build(), W3C_TEST_TRACE_ID, W3C_TEST_SPAN_ID);
+    assertTraceSpanAndSampled(builder.build(), W3C_TEST_TRACE_ID, W3C_TEST_SPAN_ID, true);
   }
 
-  private void assertTraceAndSpan(Context context, String expectedTraceId, String expectedSpanId) {
+  @Test
+  public void testParsingOpenTelemetryContext() {
```
Comment (gkevinzheng) [inline google-cloud-logging/src/test/java/com/google/cloud/logging/ContextTest.java:165]: Have you tested the case where you have an OpenTelemetry span and a request with either a Traceparent or XCTC header?

File: google-cloud-logging/src/main/java/com/google/cloud/logging/Context.java [hunk 2]
```diff
@@ -34,22 +35,27 @@ public class Context {
   private static final Pattern W3C_TRACE_CONTEXT_FORMAT =
       Pattern.compile(
           "^00-(?!00000000000000000000000000000000)[0-9a-f]{32}-(?!0000000000000000)[0-9a-f]{16}-[0-9a-f]{2}$");
+  // Trace sampled flag for bit masking
+  // see https://www.w3.org/TR/trace-context/#trace-flags for details
+  private static final byte FLAG_SAMPLED = 1; // 00000001
```
[the remark this answers is not in the document]
Comment (cindy-peng) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/Context.java:40]: I followed the official example here so I think we should be good? https://www.w3.org/TR/trace-context/#trace-flags

File: google-cloud-logging/src/main/java/com/google/cloud/logging/ContextHandler.java [hunk 3]
```diff
@@ -18,7 +18,17 @@
...
+  public enum ContextPriority {
+    NO_INPUT,
+    XCLOUD_HEADER,
+    W3_HEADER,
+    OTEL_EXTRACTED
+  }
+
   private static final ThreadLocal<Context> contextHolder = initContextHolder();
+  private static final ThreadLocal<ContextPriority> currentPriority =
+      new ThreadLocal<ContextPriority>();
```
[the remark this answers is not in the document]
Comment (cindy-peng) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/ContextHandler.java]: Good feedback, done.

File: google-cloud-logging/src/test/java/com/google/cloud/logging/ContextTest.java (hunk 1 above)
Comment (cindy-peng) [reply google-cloud-logging/src/test/java/com/google/cloud/logging/ContextTest.java:165]: I have added some system tests here: https://github.com/googleapis/java-logging/pull/1596/files#diff-116da1fe69768cf63bc5b434e783fc90c20d21f1a437d2ddab7d6b677f56ea8e

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java [hunk 4]
```diff
@@ -834,13 +836,23 @@ public Iterable<LogEntry> populateMetadata(
       if (resourceMetadata != null && entry.getResource() == null) {
         entityBuilder.setResource(resourceMetadata);
       }
+
+      ContextHandler contextHandler = new ContextHandler();
+      // Populate trace/span ID from OpenTelemetry span context to logging context.
+      if (Span.current().getSpanContext().isValid()) {
+        Context.Builder contextBuilder = Context.newBuilder().loadOpenTelemetryContext();
+        contextHandler.setCurrentContext(contextBuilder.build(), ContextPriority.OTEL_EXTRACTED);
```
[the remark this answers is not in the document]
Comment (cindy-peng) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:844]: Unlike other logging libraries, java-logging library doesn't call other setCurrentContext() within the library itself. It's usually delegated to customers or integration library. An example is mentioned [here](https://docs.google.com/document/d/1X-nyfGrY6SlKGh6Z3JATwdyPJvWFKE6Yk6JkRdcVjHM/edit?resourcekey=0-69nEDr2vW96sAHiRgEj2UA&tab=t.0#heading=h.2ybpfmxfphny) (2nd approach to set trace for log entries): 

[RequestContextFilter](https://github.com/googleapis/java-logging-servlet-initializer/blob/64e246dd9f86c361d9fdfe114ea87f94c343c1a6/src/main/java/com/google/cloud/logging/servlet/RequestContextFilter.java#L50) in [java-logging-servlet-initializer](https://github.com/googleapis/java-logging-servlet-initializer/tree/main) repository will call  setCurrentContext () and load tracing context from cloud tracing or W3C headers.

And we do have a plan mentioned [here](https://docs.google.com/document/d/1X-nyfGrY6SlKGh6Z3JATwdyPJvWFKE6Yk6JkRdcVjHM/edit?resourcekey=0-69nEDr2vW96sAHiRgEj2UA&tab=t.0#heading=h.2eu80kxqc864) to switch to the new setCurrentContext(Context context, ContextPriority priority) in [java-logging-servlet-initializer](https://github.com/googleapis/java-logging-servlet-initializer/tree/main).

File: google-cloud-logging/src/main/java/com/google/cloud/logging/ContextHandler.java [hunk 5]
```diff
@@ -44,6 +54,34 @@ public void setCurrentContext(Context context) {
     contextHolder.set(context);
   }
 
+  /**
+   * Sets the context based on the priority. Overrides traceId, spanId and TraceSampled if the
+   * passed priority is higher. HttpRequest values will be retrieved and combined from existing
+   * context if HttpRequest in the new context is empty .
+   */
+  public void setCurrentContext(Context context, ContextPriority priority) {
+    if ((currentPriority.get() == null || priority.compareTo(currentPriority.get()) >= 0)
```
[the remark this answers is not in the document]
Comment (cindy-peng) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/ContextHandler.java]: I have added `priority` parameter null check and remove `currentPriority ` null check, now I don't think `currentPriority ` will be `null`.

**Review by gkevinzheng — APPROVED** [review]
LGTM

**Pull request merged**